### PR TITLE
initial stream-k implementation with example

### PR DESCRIPTION
**Pull request by carlushuang**
streamk implementation with example, use `make example_gemm_xdl_streamk` to build.

- support both `atomic` and `reduction` mode for the partial acc accumulation cross wgs
- for `reduction` mode use less workspace for partial acc (compared with cutlass)
- `atomic` version is better than splitk, but `reduction` mode still need performance optimization. So now by default use `atomic` version.

### CODE REVIEW

File: example/01_gemm/run_gemm_example.inc [hunk 1]
```diff
@@ -25,12 +25,35 @@ bool run_gemm(const ProblemSize& problem_size, const ExecutionConfig& config)
             }
         };
 
+    auto f_get_default_stride = [](std::size_t row, std::size_t col, std::size_t stride, auto layout) {
+        if(stride == 0) {
+            // give a chance if stride is zero, return a defalt packed stride
```
Comment (keryell) [inline example/01_gemm/run_gemm_example.inc]: Use a spell checker
```suggestion
            // give a chance if stride is zero, return a default packed stride
```

File: include/ck/tensor_operation/gpu/grid/block_to_ctile_map.hpp [hunk 2]
```diff
@@ -634,5 +635,272 @@ struct BlockToCTileMap_3DGrid_KSplit
...
+        uint32_t m, uint32_t n, uint32_t k, uint32_t num_cu, uint32_t occupancy, uint32_t tile_swizzle_sub_m_factor = 8)
+    {
+        uint32_t num_tiles =
+            math::integer_divide_ceil(m, MPerBlock) * math::integer_divide_ceil(n, NPerBlock);
+        k_iters_per_tile = MDiv(math::integer_divide_ceil(k, KPerBlock));
+
+        // one cu can hold one wg at one time, from the whole chip's point of view
+        // if number of wg is same as num_cu, we call it 1 dispatch
+        // if number of wg is 2x num_cu, we call it 2 dispatches.
+        // one dispatch can deliever wg same as num_cu (full dispatch), or less than num_cu (partial
```
Comment (keryell) [inline include/ck/tensor_operation/gpu/grid/block_to_ctile_map.hpp]:
```suggestion
        // one dispatch can deliver wg same as num_cu (full dispatch), or less than num_cu (partial
```

File: example/01_gemm/run_gemm_example.inc (hunk 1 above)
Comment (carlushuang) [reply example/01_gemm/run_gemm_example.inc]: @keryell thanks for this review, I'll fix this problem in later commit

File: include/ck/tensor_operation/gpu/grid/block_to_ctile_map.hpp (hunk 2 above)
Comment (carlushuang) [reply include/ck/tensor_operation/gpu/grid/block_to_ctile_map.hpp]: @keryell thanks for this review, I'll fix this problem in later commit

File: library/src/utility/device_memory.cpp [hunk 3]
```diff
@@ -10,20 +10,49 @@ DeviceMem::DeviceMem(std::size_t mem_size) : mMemSize(mem_size)
...
+}
+
 void* DeviceMem::GetDeviceBuffer() const { return mpDeviceBuf; }
 
 std::size_t DeviceMem::GetBufferSize() const { return mMemSize; }
 
 void DeviceMem::ToDevice(const void* p) const
 {
-    hip_check_error(hipMemcpy(mpDeviceBuf, const_cast<void*>(p), mMemSize, hipMemcpyHostToDevice));
+    if(mpDeviceBuf)
```
Comment (asroy) [inline library/src/utility/device_memory.cpp:29]: should throw error if false

Comment (carlushuang) [reply library/src/utility/device_memory.cpp:29]: fixed

**Review by zjing14 — APPROVED** [review]

**Pull request merged**